Weather

hobleyd opened the pull request:
All,

Here is a big PR, sorry about that. This uses the blank space in the SearchBar to provide a weather lookup if required; I have implemented 2 weather services so far; OpenWeather which is global and the Australian BOM site which is strictly for AU. It will dynamically add icons in to fill the available space.

The AU site does not require an API Key, while the OpenWeather one does. I didn't want to register on your behalf, so the functionality is there to either embed it within the application, or ask the user to specify in the preferences.

I think I have tested most cases; but it is pretty complicated, so I may have missed something. Single clicking on the icons will switch between daily and hourly (AU) or 3-hourly (OpenWeather). Double clicking will open the related Weather app for a full forecast rather than the limited one this provides. Long clicking allows you to select a stored location, location services, or to find a new location from that service. I haven't implemented the ability to remove a stored location as I need to think about the UI for this.

Not sure if this is a problem for you or not, but I used forecastie from F-Droid as the open source preferred OpenWeather app, but this is not available in the Google Store. Not an issue if you have F-Droid installed, but fyi.

Lastly, the Location Services uses the new Google Play APIs - if this is a problem I can revert to the AOSP version, let me know.

Happy to discuss thoughts and concerns as always.

Cheers,
David

gsantner commented on app/build.gradle:
```diff
@@ -99,6 +99,8 @@ dependencies {
     implementation 'com.android.support.constraint:constraint-layout:1.1.0'
     implementation 'com.jakewharton.threetenabp:threetenabp:1.2.1'
+    implementation 'com.android.volley:volley:1.1.0'
+    implementation 'com.google.android.gms:play-services-location:16.0.0'
```
Google Play services are a absolute no-go. This makes it impossible to release on F-Droid.

hobleyd replied:
Yeah - thought this might be a problem. I'll rectify.

hobleyd replied:
OK, replaced with AOSP equivalent. Sorry about that.

gsantner commented on app/src/main/java/com/benny/openlauncher/weather/BOMWeatherService.java:
```diff
@@ -0,0 +1,186 @@
+ * Thanks to https://github.com/tonyallan/weather-au for the basics for this class.
+ */
+
+public class BOMWeatherService extends WeatherService {
```
I don't know if or not Weather stuff is enabled by default or not. But ensure that by default everything is disabled AND zero internet connections are done till enabling weather.

hobleyd replied:
By default everything is disabled until the user explicitly selects a Weather Service. Then they'll have to accept the Location permissions (if required) and requests will start to be made.

gsantner commented:
@dkanada 
I'm unsure if it's worth to add location/gps to OpenLauncher - from both battery life and privacy aspects.
Whats your thoughts on this?

hobleyd commented:
All, 

If you request Location Services, it only uses coarse network location so minimal impact to battery life; I don't actually use the fine network location (GPS) as that level of detail is not required for weather forecasts. Additionally, it only queries location once per hour, or if you move more than 25klms, so should not have a big impact there either.

However, Location services are not required at all either unless you want them. You can specify a town or postcode manually and it will work without using location services at all. All at user discretion.

I travel a lot for work, so very helpful for me.

gsantner commented:
OK I see, the rough location should be fine. Travel: Understandable use case :D.

As of code: Does it work when you turn location on/off in settings? Or remove permission in e.g. settings? Make sure this whole does no crashes on Launcher when.

To achieve that:
* Do suggest to move as much as possible to some weather java package or/and class. As far I can see stuff is scattered around all the place
* Rename icons to `weather_` prefix.
* Add license of weather icons. BTW is there some "public domain /cc0" set available? If you can find one at the fast, use that

dkanada commented:
I'm not too keen on a weather service either, but if the code is clean I think it's fine to merge. I'd much rather improve widget support and just offload calendars, clocks, alarms, and weather to other apps eventually. That would require one or two changes first though.

hobleyd commented:
Thanks for your comments:

In terms of turning location services off; Location Services are only requested if you have selected a Weather Service, and not specified a specific location to check (manually). If Location Services are required, it checks the permissions every time it is requested on the assumption that people are likely to say yes if they want to get the weather for a random location. If the permission gets revoked, the next time it tries to display the weather (up to an hour generally) then it will re-request the permission.

I could make this more explicit if you like by putting a specific setting in for Allowing Weather to request location services rather than inferring it from their preferences. Let me know if you would prefer this.

I have renamed the weather icons; they were downloaded from www.flaticon which has a free attribution licence which I have put into the README.md file. Is that sufficient do you think? Or would you prefer I add something in to the app somewhere? It looks reasonable based on the licence. I'll also look for CC icons.

In terms of the code breakdown, I am happy to take suggestions here. In essence, all the backend code is localised to the weather subfolder. On the basis of your comment, I have moved the UI code from the SearchBar into WeatherService albeit with tight coupling between the two classes given the UI code needs to be tied in with the SearchBar concepts. Happy to make further changes, if you have any specific requirements.

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -556,7 +607,9 @@ protected void onResume() {
         } else {
             setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
         }
+
```
unneeded change

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -567,6 +620,7 @@ protected void onDestroy() {
         unregisterReceiver(_appUpdateReceiver);
         unregisterReceiver(_shortcutReceiver);
         unregisterReceiver(_timeChangedReceiver);
+
```
unneeded change

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -506,6 +526,37 @@ protected void onStart() {
+            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
+                Tool.toast(HomeActivity.this, getString(R.string.toast_location_permission_required));
+                ActivityCompat.requestPermissions(HomeActivity.this,
+                        new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.INTERNET},
```
INTERNET is always granted, no need to request

gsantner commented on app/src/main/java/com/benny/openlauncher/util/AppSettings.java:
```diff
@@ -343,6 +352,118 @@ public boolean getAppRestartRequired() {
+    public String getWeatherAPIKey() {
+        String key = getString("pref_key__weather_service_apikey", "");
+
+        if (key.equals("")) {
```
TextUtils.isempty

gsantner commented on app/src/main/java/com/benny/openlauncher/util/AppSettings.java:
```diff
@@ -343,6 +352,118 @@ public boolean getAppRestartRequired() {
+        setString("pref_key__weather_service_apikey", apiKey);
+    }
+
+    public String getWeatherMeasurementUnits() {
```
maye isMetricUnit (default true) suits better here?

gsantner commented on app/src/main/java/com/benny/openlauncher/util/AppSettings.java:
```diff
@@ -343,6 +352,118 @@ public boolean getAppRestartRequired() {
+        ArrayList<WeatherLocation> storedLocations = getWeatherLocations();
+
+        if (storedLocations.contains(loc)) {
+            LOG.debug("Removing location {} from favourites.", loc);
```
comment / remove, not needed anymroe?

hobleyd replied:
This will be required; I need to decide on a UI interaction to remove stored locations; I haven't decided what that should be currently. Would you like me to remove this until it is needed?

gsantner replied:
I was talking about the log.debug not functionality

hobleyd replied:
Sorry - misunderstood. Fixed.

gsantner commented on app/src/main/java/com/benny/openlauncher/util/IconPackHelper.java:
```diff
@@ -180,4 +182,18 @@ private static Matrix getResizedMatrix(Bitmap bm, int newHeight, int newWidth) {
         return matrix;
     }
+
+    public static Drawable resizeDrawable(Drawable d, int newWidth, int newHeight) {
```
I think theres a method already in ContextUtils, take a look if you find something there

hobleyd replied:
ContextUtils only contains utilities to modify bitmaps. While this one is Drawable focused. Would you like me to move it into ContextUtils?

gsantner commented on app/src/main/java/com/benny/openlauncher/weather/BOMWeatherService.java:
```diff
@@ -0,0 +1,193 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/*
```
License?

hobleyd replied:
I wrote all the code; I did refer to the python project which documented the API that the BOM uses; but obviously no code was used.

Happy to put a licence in here if you like, but other files I spot checked had nothing something specific in there; happy to be guided by you here.

gsantner commented on app/src/main/java/com/benny/openlauncher/weather/OpenWeatherService.java:
```diff
@@ -0,0 +1,219 @@
+ * download it!
+ */
+public class OpenWeatherService extends WeatherService {
+    private static Logger LOG = LoggerFactory.getLogger("WeatherService");
```
Do we really need multiple loggers? theres already another one..plus builtin Log.d

hobleyd replied:
The beauty with slf4j is that the logging can be turned on and off again at runtime and has minor performance improvements as it doesn't need to resolve parameters if the logging is turned off. 

If you feel strongly about it, I can remove it, but I think it is a much better logging system than the existing ones. Let me know.

gsantner replied:
Preferably use one slf4j instance instead of multiple ones per class

hobleyd replied:
Done.

gsantner commented on app/src/main/res/values/strings.xml:
```diff
@@ -242,11 +256,26 @@
     <string name="widget">Widget</string>
 
+    <!-- Weather Settings -->
+    <string name="weather_service_au_bom">Australian BOM</string>
```
```
    <string name="weather_service_au_bom">Australian BOM</string>
    <string name="weather_service_openweather">OpenWeather</string>
    <string name="weather_service_1_hour">+1h</string>
    <string name="weather_service_3_hours">+3h</string>
    <string name="weather_service_1_day">+1d</string>
```

are not translatable to my knowledge. move them to the other file and mark as untrasnalatable

hobleyd replied:
Moved.

gsantner commented on app/src/main/res/values/strings_not_translatable.xml:
```diff
@@ -12,6 +12,7 @@
     <string name="pref_default__date_bar_date_format_custom_1" translatable="false"><![CDATA[EEEE\', \'dd\'.\']]></string>
     <string name="pref_default__date_bar_date_format_custom_2" translatable="false"><![CDATA[MMMM yyyy\', KW \'ww]]></string>
     <string name="pref_default__search_bar_base_uri" translatable="false"><![CDATA[https://duckduckgo.com/?kae=c&k1=-1&kaj=m&kv=l&kp=-1&q={query}]]></string>
+    <string name="pref_default__weather_service_city" translatable="false">Enter Postcode</string>
```
this is translatable

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -506,6 +526,37 @@ protected void onStart() {
+
+    public boolean checkLocationPermissions() {
+        if (checkCallingOrSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
+            && checkCallingOrSelfPermission(Manifest.permission.INTERNET) == PackageManager.PERMISSION_GRANTED) {
```
INTERNET

hobleyd replied:
Done. Sorry - didn't notice that as well.

gsantner commented on app/src/main/res/values/strings.xml:
```diff
@@ -122,6 +122,17 @@
+    <string name="pref_title__weather_service_apikey">API Key</string>
+    <string name="pref_summary__weather_service_apikey">Specify the API Key for the Weather Service (if required).</string>
+    <string name="pref_title__weather_service_units">Metric?</string>
+    <string name="pref_summary__weather_service_units">Select if you want Metric measurements, or deselect for Imperial. OpenWeather only currently.</string>
```
OpenWeatherMap?

hobleyd replied:
Changed name(s)!

gsantner commented:
I made another review @hobleyd 
Sorry if annoying, but lets get it into good shape :-)

hobleyd commented:
Not a problem at all @gsantner I want to do this properly. 

I think I have fixed up all the things you mentioned previously - I can't see any new requests. Have I missed anything?

gsantner commented on app/src/main/java/com/benny/openlauncher/util/AppSettings.java:
```diff
@@ -343,6 +353,115 @@ public boolean getAppRestartRequired() {
+        }
+    }
+
+    private String serialiseWeatherLocations(ArrayList<WeatherLocation> locations) {
```
serialise -> serialize typo (multiple places here)

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -67,14 +74,20 @@
 import java.util.Set;
 
 public final class HomeActivity extends Activity implements OnDesktopEditListener {
+    public static Logger LOG = LoggerFactory.getLogger("HomeActivity");
```
Can you remove that one as well, and use the one from AppSettings?

gsantner commented on app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:
```diff
@@ -506,6 +526,36 @@ protected void onStart() {
+                Tool.toast(HomeActivity.this, getString(R.string.toast_location_permission_required));
+                ActivityCompat.requestPermissions(HomeActivity.this,
+                        new String[]{Manifest.permission.ACCESS_COARSE_LOCATION},
+                        REQUEST_PERMISSION_NETWORK);
```
REQUEST_PERMISSION_NETWORK

hobleyd replied:
That's the constant I defined for the OnActivityResult. Admittedly not used, but it does require an id when calling requestPermission.

I have renamed it to REQUEST_PERMISSION_COARSE_LOCATION to make it more obvious.

gsantner commented on app/src/main/java/com/benny/openlauncher/util/AppSettings.java:
```diff
@@ -343,6 +353,115 @@ public boolean getAppRestartRequired() {
+            }
+        }
+
+        return "";
```
remove empty line above and below

gsantner commented on app/src/main/java/com/benny/openlauncher/util/IconPackHelper.java:
```diff
@@ -180,4 +182,18 @@ private static Matrix getResizedMatrix(Bitmap bm, int newHeight, int newWidth) {
+            return d;
+        }
+
+        Bitmap b = ((BitmapDrawable) d).getBitmap();
```
you can remove the empty lines

gsantner commented on app/src/main/java/com/benny/openlauncher/weather/WeatherResult.java:
```diff
@@ -0,0 +1,69 @@
+     * Complicated, but width and height could change if the screen is rotated.
+     */
+    public Drawable getWeatherIcon(int resourceId, int size) {
+        HashMap<Integer, Drawable> heightMap;
```
Memory and crash-safe wise, wouldn't it be better to save the drawable as integer-ressource ID instead of the object?

hobleyd replied:
It's a good question - my thinking was that the Drawables are not huge and not regenerating them means it saves CPU cycles (& hence battery life) which I thought would be more important.

If you feel strongly about it, I can change it.

gsantner replied:
I just know from my other apps thats really easy to get installed on lowest end devices. You can get a outofmemoryexception (crashing the whole app) pretty fast.

leaving this up to you for now

hobleyd commented:
@gsantner , I have made all those changes (with one query). I am flying back to Australia tomorrow morning so will be out of action for a day (just in case you are wondering why I don't respond to updates).

hobleyd commented:
@gsantner , my humblest apologies. I should have regression tested this more fully obviously. I believe I have everything working as you would expect now.

I have moved out the Weather City from the preferences as I can't easily put a drop down on a the v7 support preferences; replaced it with an explicit UseLocationServices preference; if UseLocationServices is not set and the Weather Service has been requested, it will prompt for a city which is searched via the relevant Weather Service.

gsantner commented:
No problem at all!

After all that's why I'm doing this reviews and help to improve it!

hobleyd commented:
@gsantner, One question - have you signed up with OpenWeatherMap for a default API key? Or would you expect people to do that themselves? If the latter, I'll need to do some more work with error handling for that scenario. I'll get to that in the next day or two.

gsantner commented:
I didn't follow, whats the status?

hobleyd commented:
Sorry for the delay - I have been a bit distracted by work and a couple of other projects; I actually have Android Studio open to this today and will be looking at it this week.

gsantner commented:
What is the status on this?

hobleyd commented:
I'm trying to track down an irritating layout bug which I am having difficulty reproducing.

gsantner commented:
Closing due no response and inactivity.
